lib/teleterm/apiserver/apiserver.go

```diff
@@ -78,7 +86,63 @@ func New(cfg Config) (*APIServer, error) {
 
 	api.RegisterTerminalServiceServer(grpcServer, serviceHandler)
 
-	return &APIServer{cfg, ls, grpcServer}, nil
+	apiServer := &APIServer{
+		Config:     cfg,
+		closeC:     make(chan struct{}),
+		ls:         ls,
+		grpcServer: grpcServer,
+	}
+
+	go func() {
+		err := apiServer.createAndInjectTshdEventsClient(startupServiceHandler, shouldUseMTLS, tshdKeyPair)
+		if err != nil {
+			cfg.Log.WithError(err).Error("Could not create tshd events client")
+		}
+	}()
+
+	return apiServer, nil
+}
+
+// createAndInjectTshdEventsClient is executed in a goroutine from apiserver.New.
+// It waits for the Electron app to call the startup service to provide the tshd events server
+// address. Then it initializes the client for that server and injects it into daemon.Server.
+func (s *APIServer) createAndInjectTshdEventsClient(startupServiceHandler *startuphandler.Handler, shouldUseMTLS bool, tshdKeyPair tls.Certificate) error {
+	s.Log.Info("Waiting for tshd events server address")
+
+	select {
+	case <-s.closeC:
+		return trace.Errorf("waiting for tshd events server address aborted because APIServer is closing")
+	case <-startupServiceHandler.WaitForTshdEventsServerAddressC:
+	}
+
+	tshdEventsServerAddress := startupServiceHandler.TshdEventsServerAddress()
```

Comment:
I'm not sure if that's the best way to approach this.

Right now I have a channel, I wait for it to be closed and once it's closed I know that there's a string that will be available on another field on the struct.

I could just send this string over a buffered channel instead. This initially felt wrong to me because then the value can be read from the channel only once (vs reading it as many times as we want through a struct field). But in this scenario we need it only once anyway so… I guess it'd be better to simplify this?

Reply:
@gzdunek Pointed out that because `startuphandler.Handler.ResolveTshdEventsServerAddress` returns an error when called more than once during the lifetime of the tsh daemon, during development you can no longer use `Cmd + R` to refresh the app.

I think it points at the underlying problem with how I designed the backend to work.

The only reason `WaitForTshdEventsServerAddressC` exists is because I want to be notified when the frontend app calls `ResolveTshdEventsServerAddress` because I want to set up a client for tshd events immediately after.

Instead of spinning up a goroutine and having a separate channel, I can just make it so that each call to `ResolveTshdEventsServerAddress` kicks up the whole process of setting up a client. This will make refreshing work in development. In production, the endpoint would still be called only once.

I believe this might simplify the code overall, so I'll see what I can do here.